### PR TITLE
Modify List to Map of pooled redis message in FixedRedisMessagePool

### PR DESCRIPTION
**Pull request by jjz921024**
Motivation:

For easy to get pooled message, add two type enum indicate Redis message.
Use an enum as the key to easy get pooled message.

Modifications:

Modify pooled collections from List to Map in FixedRedisMessagePool,
Use an enum as the key to easy get pooled message.

Result:

Users can get pooled message from map by enum instead of the whole string

### CODE REVIEW

File: codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java [hunk 1]
```diff
@@ -118,8 +128,8 @@ private FixedRedisMessagePool() {
     }
 
     @Override
-    public SimpleStringRedisMessage getSimpleString(String content) {
-        return stringToSimpleStrings.get(content);
+    public SimpleStringRedisMessage getSimpleString(RedisReplyKey key) {
```
Comment (hyperxpro) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java:131]: API Breakage

File: codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java [hunk 2]
```diff
@@ -128,8 +138,8 @@ public SimpleStringRedisMessage getSimpleString(ByteBuf content) {
     }
 
     @Override
-    public ErrorRedisMessage getError(String content) {
-        return stringToErrors.get(content);
+    public ErrorRedisMessage getError(RedisErrorKey key) {
```
Comment (hyperxpro) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java:141]: API Breakage

File: codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java [hunk 3]
```diff
@@ -26,19 +26,19 @@
 public interface RedisMessagePool {
 
     /**
-     * Returns {@link SimpleStringRedisMessage} for given {@code content}. Returns {@code null} it does not exist.
+     * Returns {@link SimpleStringRedisMessage} for given {@code RedisReplyKey}. Returns {@code null} it does not exist.
      */
-    SimpleStringRedisMessage getSimpleString(String content);
+    SimpleStringRedisMessage getSimpleString(FixedRedisMessagePool.RedisReplyKey key);
```
Comment (hyperxpro) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java:31]: API Breakage

File: codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java [hunk 4]
```diff
@@ -26,19 +26,19 @@
...
      * Returns {@link SimpleStringRedisMessage} for given {@code content}. Returns {@code null} it does not exist.
      */
     SimpleStringRedisMessage getSimpleString(ByteBuf content);
 
     /**
-     * Returns {@link ErrorRedisMessage} for given {@code content}. Returns {@code null} it does not exist.
+     * Returns {@link ErrorRedisMessage} for given {@code RedisErrorKey}. Returns {@code null} it does not exist.
      */
-    ErrorRedisMessage getError(String content);
+    ErrorRedisMessage getError(FixedRedisMessagePool.RedisErrorKey key);
```
Comment (hyperxpro) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java:41]: API Breakage

**Review by hyperxpro — CHANGES_REQUESTED** [review]
Well, this change will break applications.

**Comment by normanmaurer** [comment]
@jjz921024 as @hyperxpro said we can't break the API...

**Comment by jjz921024** [comment]
> @jjz921024 as @hyperxpro said we can't break the API...

Sorry, I haven't thought about that.
Can i add two method for getRedisMessage by enum.
In the meantime, don't modify former API.
`
   SimpleStringRedisMessage getSimpleString(FixedRedisMessagePool.RedisReplyKey key);
   ErrorRedisMessage getError(FixedRedisMessagePool.RedisErrorKey key);
`
Because i think, this will useful for user and can avoid `hard coding` in their program

**Comment by hyperxpro** [comment]
> > @jjz921024 as @hyperxpro said we can't break the API...
> 
> Sorry, I haven't thought about that.
> Can i add two method for getRedisMessage by enum.
> In the meantime, don't modify former API.
> `SimpleStringRedisMessage getSimpleString(FixedRedisMessagePool.RedisReplyKey key); ErrorRedisMessage getError(FixedRedisMessagePool.RedisErrorKey key);`
> Because i think, this will useful for user and can avoid `hard coding` in their program

Commit the changes and we'll review it.

**Comment by jjz921024** [comment]
@hyperxpro Changes had commit.

**Comment by hyperxpro** [comment]
I am talking about changes post my review. Have you fixed API breakage?

**Comment by jjz921024** [comment]
> I am talking about changes post my review. Have you fixed API breakage?

@hyperxpro I recover api of getSimpleString & getError by String in second commit，keep the exist api  unchanged.

File: codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java (hunk 4 above)
Comment (normanmaurer) [reply codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java:51]: @jjz921024 you can't add new methods to an interface as this is an API breakage.

Comment (jjz921024) [reply codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java:51]: @normanmaurer @hyperxpro Sorry, I didn't realize that. So, I can only add methods to the impl class of this interface?
I will fixed it

Comment (jjz921024) [reply codec-redis/src/main/java/io/netty/handler/codec/redis/RedisMessagePool.java]: > Can you add javadocs to the two public methods ?

@normanmaurer yes, I've added javadocs to those two methods.

File: codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java [hunk 5]
```diff
@@ -122,6 +138,13 @@ public SimpleStringRedisMessage getSimpleString(String content) {
         return stringToSimpleStrings.get(content);
     }
 
+    /**
+     * Returns {@link SimpleStringRedisMessage} for given {@link RedisReplyKey}. Returns {@code null} it does not exist.
```
Comment (normanmaurer) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java:142]:
```suggestion
     * Returns {@link SimpleStringRedisMessage} for the given {@link RedisReplyKey} or {@code null} if it does not exist.
```

File: codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java [hunk 6]
```diff
@@ -132,6 +155,13 @@ public ErrorRedisMessage getError(String content) {
         return stringToErrors.get(content);
     }
 
+    /**
+     * Returns {@link ErrorRedisMessage} for given {@link RedisErrorKey}. Returns {@code null} it does not exist.
```
Comment (normanmaurer) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java:159]:
```suggestion
     * Returns {@link ErrorRedisMessage} for the given {@link RedisErrorKey} or {@code null} if it does not exist.
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Can you also add some unit tests please ?

**Comment by jjz921024** [comment]
@normanmaurer yes, i have added some unit test case for FixedRedisMessagePool

File: codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
+package io.netty.handler.codec.redis;
+
+import org.junit.Test;
```
Comment (normanmaurer) [inline codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java:3]: Please use junit 5

File: codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
+package io.netty.handler.codec.redis;
```
Comment (normanmaurer) [inline codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java:1]: Missing license header

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    public void shouldReturnDifferentMessage() {
+        FixedRedisMessagePool pool = FixedRedisMessagePool.INSTANCE;
+
+        SimpleStringRedisMessage okMessage = pool.getSimpleString(FixedRedisMessagePool.RedisReplyKey.OK);
+        SimpleStringRedisMessage pongMessage = pool.getSimpleString(FixedRedisMessagePool.RedisReplyKey.PONG);
+
+        assertThat(okMessage, is(not(pongMessage)));
+    }
+
+}
```
Comment (hyperxpro) [inline codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java:60]: Add empty line here.

File: codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2016 The Netty Project
```
Comment (normanmaurer) [inline codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java:2]: nit: 2021

**Comment by normanmaurer** [comment]
@jjz921024 please fix the checkstyle error:

```
Error:  /home/runner/work/netty/netty/codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java:1: File does not end with a newline. [NewlineAtEndOfFile]
2513
Audit done.
```

File: codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java (hunk 10 above)
Comment (normanmaurer) [inline codec-redis/src/test/java/io/netty/handler/codec/redis/FixedRedisMessagePoolTest.java:2]:
```suggestion
 * Copyright 2021 The Netty Project
```

File: codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java [hunk 11]
```diff
@@ -152,3 +184,4 @@ public IntegerRedisMessage getInteger(ByteBuf content) {
         return longToByteBufs.get(value);
     }
 }
+
```
Comment (trustin) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java:187]: nit: Could be removed

**Review by trustin — APPROVED** [review]
Approved with one nit

File: codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java [hunk 12]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016 The Netty Project
+ * Copyright 2021 The Netty Project
```
Comment (normanmaurer) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java:2]:
```suggestion
 * Copyright 2016 The Netty Project
```

We don't change copyright years

File: codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java (hunk 11 above)
Comment (normanmaurer) [inline codec-redis/src/main/java/io/netty/handler/codec/redis/FixedRedisMessagePool.java:187]:
```suggestion
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by normanmaurer** [comment]
@jjz921024 thanks a lot !

**Pull request merged**